File: ArithmeticSlices_DP.java
// Time Complexity : O(n) --> where n is the length of input array
// Space Complexity : O(1)
// Did this code successfully run on Leetcode (413): Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

class Solution {
    public int numberOfArithmeticSlices(int[] A) {
        // edge case
        if (A == null || A.length < 3) return 0;
        
        int prev = 0;
        int sum = 0;
        for (int i = 2; i < A.length; i++) {
            if (A[i-1] - A[i-2] == A[i] - A[i-1]) {
                prev += 1;
                sum += prev;
            }
            else prev = 0;
        }
        return sum;
    }
}

File: arithmetic_slices_dp.py
# Time Complexity : O(N) where N is length of the nums array
# Space Complexity : O(N) where N is length of the nums array
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        count = 0
        n = len(nums)
        dp = [0 for _ in range(len(nums))]
        diff  = nums[n-1] - nums[n-2]
        for i in range(n-3, -1, -1):
            if  nums[i+1] - nums[i] == diff:
                dp[i] = dp[i+1] + 1
                count += dp[i]
            else:
                diff  = nums[i+1] - nums[i]
                
        return count